File: php/src/Api/Destination.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Api;

use Svix\Exception\ApiException;
use Svix\Models\DestinationIn;
use Svix\Models\DestinationOut;
use Svix\Models\DestinationPatch;
use Svix\Models\ListResponseDestinationOut;
use Svix\Request\SvixHttpClient;

class Destination
{
    public DestinationTransformation $transformation;

    public function __construct(
        private readonly SvixHttpClient $client,
    ) {
        $this->transformation = new DestinationTransformation($client);
    }

    /**
     * List of all the application's destinations.
     *
     * @throws ApiException
     */
    public function list(
        string $appId,
        DestinationListOptions $options = new DestinationListOptions(),
    ): ListResponseDestinationOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/destination");
        $request->setQueryParam('limit', $options->limit);
        $request->setQueryParam('iterator', $options->iterator);
        $request->setQueryParam('order', $options->order);
        $res = $this->client->send($request);

        return ListResponseDestinationOut::fromJson($res);
    }

    /**
     * Creates a new destination.
     *
     * @throws ApiException
     */
    public function create(
        string $appId,
        DestinationIn $destinationIn,
        DestinationCreateOptions $options = new DestinationCreateOptions(),
    ): DestinationOut {
        $request = $this->client->newReq('POST', "/api/v1/app/{$appId}/destination");
        $request->setHeaderParam('idempotency-key', $options->idempotencyKey);
        $request->setBody(json_encode($destinationIn));
        $res = $this->client->send($request);

        return DestinationOut::fromJson($res);
    }

    /**
     * Get a destination by id or uid.
     *
     * @throws ApiException
     */
    public function get(
        string $appId,
        string $destinationId,
    ): DestinationOut {
        $request = $this->client->newReq('GET', "/api/v1/app/{$appId}/destination/{$destinationId}");
        $res = $this->client->send($request);

        return DestinationOut::fromJson($res);
    }

    /**
     * Create or update a destination.
     *
     * @throws ApiException
     */
    public function upsert(
        string $appId,
        string $destinationId,
        DestinationIn $destinationIn,
    ): DestinationOut {
        $request = $this->client->newReq('PUT', "/api/v1/app/{$appId}/destination/{$destinationId}");
        $request->setBody(json_encode($destinationIn));
        $res = $this->client->send($request);

        return DestinationOut::fromJson($res);
    }

    /**
     * Delete a destination.
     *
     * @throws ApiException
     */
    public function delete(
        string $appId,
        string $destinationId,
    ): void {
        $request = $this->client->newReq('DELETE', "/api/v1/app/{$appId}/destination/{$destinationId}");
        $res = $this->client->sendNoResponseBody($request);
    }

    /**
     * Partially update a destination.
     *
     * @throws ApiException
     */
    public function patch(
        string $appId,
        string $destinationId,
        DestinationPatch $destinationPatch,
    ): DestinationOut {
        $request = $this->client->newReq('PATCH', "/api/v1/app/{$appId}/destination/{$destinationId}");
        $request->setBody(json_encode($destinationPatch));
        $res = $this->client->send($request);

        return DestinationOut::fromJson($res);
    }
}

File: kotlin/lib/src/main/kotlin/api/Destination.kt
// this file is @generated
package com.svix.kotlin.api

import com.svix.kotlin.SvixHttpClient
import com.svix.kotlin.models.DestinationIn
import com.svix.kotlin.models.DestinationOut
import com.svix.kotlin.models.DestinationPatch
import com.svix.kotlin.models.ListResponseDestinationOut
import com.svix.kotlin.models.Ordering
import com.svix.kotlin.serializeQueryParam
import okhttp3.Headers

data class DestinationListOptions(
    /** Limit the number of returned items */
    val limit: ULong? = null,
    /** The iterator returned from a prior invocation */
    val iterator: String? = null,
    /** The sorting order of the returned items */
    val order: Ordering? = null,
)

data class DestinationCreateOptions(val idempotencyKey: String? = null)

class Destination(private val client: SvixHttpClient) {
    val transformation: DestinationTransformation = DestinationTransformation(client)

    /** List of all the application's destinations. */
    suspend fun list(
        appId: String,
        options: DestinationListOptions = DestinationListOptions(),
    ): ListResponseDestinationOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/destination")
        options.limit?.let { url.addQueryParameter("limit", serializeQueryParam(it)) }
        options.iterator?.let { url.addQueryParameter("iterator", it) }
        options.order?.let { url.addQueryParameter("order", serializeQueryParam(it)) }
        return client.executeRequest<Any, ListResponseDestinationOut>("GET", url.build())
    }

    /** Creates a new destination. */
    suspend fun create(
        appId: String,
        destinationIn: DestinationIn,
        options: DestinationCreateOptions = DestinationCreateOptions(),
    ): DestinationOut {
        val url = client.newUrlBuilder().encodedPath("/api/v1/app/$appId/destination")
        val headers = Headers.Builder()
        options.idempotencyKey?.let { headers.add("idempotency-key", it) }

        return client.executeRequest<DestinationIn, DestinationOut>(
            "POST",
            url.build(),
            headers = headers.build(),
            reqBody = destinationIn,
        )
    }

    /** Get a destination by id or uid. */
    suspend fun get(appId: String, destinationId: String): DestinationOut {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/destination/$destinationId")
        return client.executeRequest<Any, DestinationOut>("GET", url.build())
    }

    /** Create or update a destination. */
    suspend fun upsert(
        appId: String,
        destinationId: String,
        destinationIn: DestinationIn,
    ): DestinationOut {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/destination/$destinationId")

        return client.executeRequest<DestinationIn, DestinationOut>(
            "PUT",
            url.build(),
            reqBody = destinationIn,
        )
    }

    /** Delete a destination. */
    suspend fun delete(appId: String, destinationId: String) {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/destination/$destinationId")
        client.executeRequest<Any, Boolean>("DELETE", url.build())
    }

    /** Partially update a destination. */
    suspend fun patch(
        appId: String,
        destinationId: String,
        destinationPatch: DestinationPatch,
    ): DestinationOut {
        val url =
            client.newUrlBuilder().encodedPath("/api/v1/app/$appId/destination/$destinationId")

        return client.executeRequest<DestinationPatch, DestinationOut>(
            "PATCH",
            url.build(),
            reqBody = destinationPatch,
        )
    }
}
